### PR TITLE
Lower `tt.load` and `tt.store` to gather/scatter memory accessing when Intel Block IO is not supported.

### PR DESCRIPTION
**Pull request by chengjunlu**
Intel Block IO HW only supports a contiguous memory accessing in Triton.
To lower the `tt.load` with block pointer by the gather/scatter memory accessing if the Block IO is not supported.

### CODE REVIEW

**Comment by chengjunlu** [comment]
I will add more LIT test to this PR.

File: third_party/intel/lib/TritonIntelGPUToLLVM/LoadStoreOpToLLVM.cpp [hunk 1]
```diff
@@ -705,56 +796,69 @@ struct LoadOpConversion
...
+      }
+    } else {
+      // original values
+      Value ptr = op.getPtr();
+      Value other = op.getOther();
+      Value mask = op.getMask();
+
+      Value llPtr = adaptor.getPtr();
+      Value llMask = adaptor.getMask();
+      Value llOther = adaptor.getOther();
```
Comment (FMarno) [inline third_party/intel/lib/TritonIntelGPUToLLVM/LoadStoreOpToLLVM.cpp:834]: Why do you need ptr, other, and mask twice?

File: third_party/intel/lib/TritonIntelGPUToLLVM/LoadStoreOpToLLVM.cpp [hunk 2]
```diff
@@ -705,56 +796,69 @@ struct LoadOpConversion
...
-      otherIsSplatConstInt = true;
-      splatVal = constAttr.getSplatValue<APInt>().getSExtValue();
-    }
     SmallVector<Value> otherElems;
-    if (other) {
-      otherElems = unpackLLElements(loc, llOther, rewriter);
+
+    // adaptor values
+    if (isTensorPointerType(op.getPtr().getType())) {
+      if (rewriteTensorPointerLoad(op, adaptor, rewriter).failed()) {
```
Comment (FMarno) [inline third_party/intel/lib/TritonIntelGPUToLLVM/LoadStoreOpToLLVM.cpp:814]: I think the style guide suggests flipping this around for the early return in the success case

File: third_party/intel/lib/TritonIntelGPUToLLVM/LoadStoreOpToLLVM.cpp [hunk 3]
```diff
@@ -996,46 +1100,60 @@ struct StoreOpConversion
...
+
     auto valueTy = value.getType();
     Type valueElemTy =
         typeConverter->convertType(getElementTypeOrSelf(valueTy));
+    SmallVector<Value> ptrElems;
+    SmallVector<Value> maskElems;
+    unsigned vec = 1;
+
+    if (isTensorPointerType(ptr.getType())) {
+      if (rewriteTensorPointerStore(op, adaptor, rewriter).failed()) {
```
Comment (FMarno) [inline third_party/intel/lib/TritonIntelGPUToLLVM/LoadStoreOpToLLVM.cpp:1117]: early return on success

File: third_party/intel/lib/TritonIntelGPUToLLVM/LoadStoreOpToLLVM.cpp [hunk 4]
```diff
@@ -996,46 +1100,60 @@ struct StoreOpConversion
...
-      vec = std::min(vec, maskAlign);
+        unsigned maskAlign = getMaskAlignment(mask);
+        vec = std::min(vec, maskAlign);
+      }
     }
 
+    Value llValue = adaptor.getValue();
+    auto valueElems = unpackLLElements(loc, llValue, rewriter);
+    assert(ptrElems.size() == valueElems.size());
+    assert(!maskElems.size() || valueElems.size() == maskElems.size());
```
Comment (FMarno) [inline third_party/intel/lib/TritonIntelGPUToLLVM/LoadStoreOpToLLVM.cpp:1149]: add assert messages

**Comment by mfrancepillois** [comment]
In the `RewriteTensorPointer` pass we check for the divisibility (see https://github.com/intel/intel-xpu-backend-for-triton/blob/b040b8eb1f587992a8d9663c1d45dec5e47c1a16/third_party/intel/lib/TritonIntelGPUTransforms/RewriteTensorPointer.cpp#L60 ) and the contiguity (https://github.com/intel/intel-xpu-backend-for-triton/blob/b040b8eb1f587992a8d9663c1d45dec5e47c1a16/third_party/intel/lib/TritonIntelGPUTransforms/RewriteTensorPointer.cpp#L64-L69) of the block pointers. 
Maybe, I missed it, but I couldn't find these checks in `rewriteTensorPointerLoad` (https://github.com/intel/intel-xpu-backend-for-triton/blob/b040b8eb1f587992a8d9663c1d45dec5e47c1a16/third_party/intel/lib/TritonIntelGPUToLLVM/LoadStoreOpToLLVM.cpp#L478-L766 ) and `rewriteTensorPointerStore` (https://github.com/intel/intel-xpu-backend-for-triton/blob/b040b8eb1f587992a8d9663c1d45dec5e47c1a16/third_party/intel/lib/TritonIntelGPUToLLVM/LoadStoreOpToLLVM.cpp#L978C3-L1098).
Is there a reason for this?

**Comment by mfrancepillois** [comment]
If the pass is no longer called, we can probably delete the source code of the pass, i.e. remove `RewriteTensorPointer.cpp`

**Comment by chengjunlu** [comment]
> In the `RewriteTensorPointer` pass we check for the divisibility (see
> 
> https://github.com/intel/intel-xpu-backend-for-triton/blob/b040b8eb1f587992a8d9663c1d45dec5e47c1a16/third_party/intel/lib/TritonIntelGPUTransforms/RewriteTensorPointer.cpp#L60
> 
> ) and the contiguity (
> https://github.com/intel/intel-xpu-backend-for-triton/blob/b040b8eb1f587992a8d9663c1d45dec5e47c1a16/third_party/intel/lib/TritonIntelGPUTransforms/RewriteTensorPointer.cpp#L64-L69
> 
> ) of the block pointers.
> Maybe, I missed it, but I couldn't find these checks in `rewriteTensorPointerLoad` (
> https://github.com/intel/intel-xpu-backend-for-triton/blob/b040b8eb1f587992a8d9663c1d45dec5e47c1a16/third_party/intel/lib/TritonIntelGPUToLLVM/LoadStoreOpToLLVM.cpp#L478-L766
> 
> ) and `rewriteTensorPointerStore` (https://github.com/intel/intel-xpu-backend-for-triton/blob/b040b8eb1f587992a8d9663c1d45dec5e47c1a16/third_party/intel/lib/TritonIntelGPUToLLVM/LoadStoreOpToLLVM.cpp#L978C3-L1098).
> Is there a reason for this?

This analysis has been moved to the new pass.
https://github.com/intel/intel-xpu-backend-for-triton/blob/e5f0e023ea81b2ebcc1cfff8267f3fa9edfd97fb/third_party/intel/lib/TritonIntelGPUTransforms/MaterializeBlockPointer.cpp#L51

It will annotate the `tt.load` with the memory contiguous information. If there is no such information in the lowering of `tt.load`, we always fallback to the gather/scatter memory accessing.

**Comment by chengjunlu** [comment]
> If the pass is no longer called, we can probably delete the source code of the pass, i.e. remove `RewriteTensorPointer.cpp`

Let's keep the optimization pipeline in this PR. I will create a new PR to clean up the `RewriteTensorPointer.cpp`.

File: third_party/intel/lib/TritonIntelGPUToLLVM/LoadStoreOpToLLVM.cpp [hunk 5]
```diff
@@ -187,6 +187,97 @@ struct LoadStoreConversionBase {
...
+  std::tuple<SmallVector<Value>, SmallVector<Value>, SmallVector<Value>>
+  convertBlockPtrToTensorOfPtr(
+      Location loc, Value blockPointerStruct, RankedTensorType tensorType,
+      Type valueElemTy, ConversionPatternRewriter &rewriter,
+      ArrayRef<int32_t> boundaryCheck = {},
+      std::optional<PaddingOption> padding = std::nullopt) const {
+
+    auto rank = tensorType.getRank();
+    unsigned blockOffset = 0 * rank, blockShape = 1 * rank,
+             blockStride = 2 * rank, blockBase = 3 * rank;
```
Comment (victor-eds) [inline third_party/intel/lib/TritonIntelGPUToLLVM/LoadStoreOpToLLVM.cpp:199]: I don't think I fully get this:
- Why is `blockOffset` always 0?
- Why the others require this product?

File: test/Conversion/intel/tritongpu_to_gen.mlir [hunk 6]
```diff
@@ -660,10 +660,10 @@ module attributes {"triton_gpu.num-ctas" = 1 : i32, "triton_gpu.num-warps" = 4 :
   tt.func @basic_store(%ptrs: tensor<256x!tt.ptr<f32>, #blocked0>, %vals: tensor<256xf32, #blocked0>, %mask: tensor<256xi1, #blocked0>) {
     // CHECK:      [[ARG0_0:%.*]] = llvm.extractvalue %arg0[0] : !llvm.struct<(ptr<1>, ptr<1>)>
     // CHECK-NEXT: [[ARG0_1:%.*]] = llvm.extractvalue %arg0[1] : !llvm.struct<(ptr<1>, ptr<1>)>
-    // CHECK-NEXT: [[ARG1_0:%.*]] = llvm.extractvalue %arg1[0] : !llvm.struct<(f32, f32)>
```
Comment (etiotto) [inline test/Conversion/intel/tritongpu_to_gen.mlir:663]: [Suggestion]: Use CHEK-DAG ?

File: test/TritonIntelGPU/blockptr_load.mlir [hunk 7]
```diff
@@ -255,3 +255,44 @@ module attributes {"triton_gpu.num-warps" = 1 : i32, "triton_gpu.threads-per-war
...
+module attributes {"triton_gpu.num-warps" = 1 : i32, "triton_gpu.threads-per-warp" = 16 : i32} {
+  // CHECK-LABEL:   llvm.func spir_kernelcc @non_contiguous_load_dot_layout
+  tt.func public @non_contiguous_load_dot_layout(%arg0: !tt.ptr<f16>, %col_stride: i64) {
+      %c64_i32 = arith.constant 64 : i32
+      %c64_i64 = arith.constant 64 : i64
+      %c1_i64 = arith.constant 1 : i64
+      %c0_i32 = arith.constant 0 : i32
+      %c32_i64 = arith.constant 32 : i64
+      %0 = tt.make_tensor_ptr %arg0, [%c64_i64, %c64_i64], [%c1_i64, %col_stride], [%c0_i32, %c0_i32] {order = array<i32: 0, 1>} : <tensor<64x16xf16, #dot_b>>
+      // CHECK-COUNT-64: llvm.load
```
Comment (etiotto) [inline test/TritonIntelGPU/blockptr_load.mlir:273]: Add the result type of the load (same at line 276

File: test/TritonIntelGPU/blockptr_load.mlir [hunk 8]
```diff
@@ -255,3 +255,44 @@ module attributes {"triton_gpu.num-warps" = 1 : i32, "triton_gpu.threads-per-war
...
+module attributes {"triton_gpu.num-warps" = 8 : i32, "triton_gpu.threads-per-warp" = 16 : i32} {
+  // CHECK-LABEL:   llvm.func spir_kernelcc @blocked_layout
+  tt.func public @blocked_layout(%arg0: !tt.ptr<f16>, %col_stride: i64) {
+      %c64_i32 = arith.constant 64 : i32
+      %c64_i64 = arith.constant 64 : i64
+      %c1_i64 = arith.constant 1 : i64
+      %c0_i32 = arith.constant 0 : i32
+      %c32_i64 = arith.constant 32 : i64
+      %21 = tt.make_tensor_ptr %arg0, [%c64_i64, %c64_i64], [%c1_i64, %col_stride], [%c0_i32, %c0_i32] {order = array<i32: 0, 1>} : <tensor<64x16xf16, #blocked>>
+      // CHECK-COUNT-32: llvm.load
```
Comment (etiotto) [inline test/TritonIntelGPU/blockptr_load.mlir:294]: Add load result type

**Comment by chengjunlu** [comment]
> Shouldn't the target branch be `llvm-target`?

The Intel Triton has migrate the default branch to the `main`.

File: third_party/intel/lib/TritonIntelGPUToLLVM/LoadStoreOpToLLVM.cpp [hunk 9]
```diff
@@ -714,56 +804,69 @@ struct LoadOpConversion
...
-    if (other) {
-      otherElems = unpackLLElements(loc, llOther, rewriter);
+
+    // adaptor values
+    if (isTensorPointerType(op.getPtr().getType())) {
+      if (rewriteTensorPointerLoad(op, adaptor, rewriter).succeeded()) {
+        return success();
+      } else {
+        // TODO: (johnlu) set the vector size > 1; Need to prove the memory is
+        // contiguous on the fast change dim when fallback to gather load.
```
Comment (whitneywhtsang) [inline third_party/intel/lib/TritonIntelGPUToLLVM/LoadStoreOpToLLVM.cpp:826]:
```suggestion
        // contiguous on the fast changing dim when fallback to gather load.
```

**Comment by whitneywhtsang** [comment]
> As the tt.load is supported to be lowered with arbitrary combination of the block pointer and layout, we can simply remove the RewriteTensorPointer.

We can remove this line from the PR description, as this PR doesn't remove RewriteTensorPointer pass.

File: third_party/intel/lib/TritonIntelGPUToLLVM/LoadStoreOpToLLVM.cpp [hunk 10]
```diff
@@ -187,6 +187,106 @@ struct LoadStoreConversionBase {
...
+    // The block pointer struct is expected to have the following layout:
+    //    Struct {
+    //      Value offset[rank];
+    //      Value shape[rank];
+    //      Value stride[rank];
+    //      Value base;
+    //    }
+    // All the values are decomposed by `unpackLLElements` into a vector.
+    // Defines the indices for the block pointer struct.
+    unsigned blockOffset = 0 * rank, blockShape = 1 * rank,
```
Comment (whitneywhtsang) [inline third_party/intel/lib/TritonIntelGPUToLLVM/LoadStoreOpToLLVM.cpp:207]:
```suggestion
    unsigned blockOffset = 0, blockShape = 1 * rank,
```

File: third_party/intel/lib/TritonIntelGPUToLLVM/LoadStoreOpToLLVM.cpp [hunk 11]
```diff
@@ -187,6 +187,106 @@ struct LoadStoreConversionBase {
...
+          }
+          return accumulate;
+        };
+
+    SmallVector<Value> ptrElems(numElems);
+    SmallVector<Value> maskElems(numElems);
+    for (unsigned i = 0; i < numElems; i++) {
+      auto index = indices[i];
+      SmallVector<Value> indicesInTensor(rank);
+      for (unsigned j = 0; j < rank; j++) {
```
Comment (whitneywhtsang) [inline third_party/intel/lib/TritonIntelGPUToLLVM/LoadStoreOpToLLVM.cpp:237]:
```suggestion
      for (unsigned j = 0; j < rank; ++j) {
```

File: third_party/intel/lib/TritonIntelGPUToLLVM/LoadStoreOpToLLVM.cpp [hunk 12]
```diff
@@ -187,6 +187,106 @@ struct LoadStoreConversionBase {
...
+            for (auto [a, b] : llvm::zip(A, B)) {
+              accumulate = linearizeFunc(a, b, accumulate);
+            }
+          }
+          return accumulate;
+        };
+
+    SmallVector<Value> ptrElems(numElems);
+    SmallVector<Value> maskElems(numElems);
+    for (unsigned i = 0; i < numElems; i++) {
```
Comment (whitneywhtsang) [inline third_party/intel/lib/TritonIntelGPUToLLVM/LoadStoreOpToLLVM.cpp:234]:
```suggestion
    for (unsigned i = 0; i < numElems; ++i) {
```

**Review by whitneywhtsang — APPROVED** [review]

**Review by victor-eds — APPROVED** [review]

**Review by FMarno — APPROVED** [review]

**Pull request merged**